core/src/main/java/io/aiven/kafka/tieredstorage/metrics/SensorProvider.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright 2023 Aiven Oy
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.aiven.kafka.tieredstorage.metrics;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.function.Supplier;
+
+import org.apache.kafka.common.MetricNameTemplate;
+import org.apache.kafka.common.metrics.MeasurableStat;
+import org.apache.kafka.common.metrics.Metrics;
+import org.apache.kafka.common.metrics.Sensor;
+
+/**
+ * Inspired by <a href="https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/consumer/internals/SensorBuilder.java">SensorBuilder</a>
+ */
+public class SensorProvider {
+    private final org.apache.kafka.common.metrics.Metrics metrics;
+
+    private final Sensor sensor;
+
+    private final boolean preexisting;
+
+    private final Map<String, String> tags;
+
+    public SensorProvider(final Metrics metrics,
+                          final String name,
+                          final Supplier<Map<String, String>> tagsSupplier) {
```

Comment:
Does it need to be a function?

Reply:
I think is a small optimization to avoid calling the supplier when sensor already exists.